### ISSUE TITLE
Error in solve.default(comp$XX): system is computationally singular

### ISSUE THREAD

**Issue by kfuku52** [khabbazian/l1ou#30]
Sorry to keep bothering you. I found another error which happens to some of my relatively large datasets (typically >150 leaves). Like the problem in [#29](https://github.com/khabbazian/l1ou/issues/29), the error can be circumvented by specifying `fixed.alpha=TRUE` in `estimate_convergent_regimes()`, although the error message is different. I would appreciate any comments or suggestion.

Input data:
[traits.txt](https://github.com/khabbazian/l1ou/files/1039445/traits.txt)
[tree.txt](https://github.com/khabbazian/l1ou/files/1039446/tree.txt)

Error message:
```
Error in solve.default(comp$XX): system is computationally singular: reciprocal condition number = 4.72733e-22
Traceback:

1. estimate_convergent_regimes(fit_ind, criterion = "AICc", method = "backward", 
 .     fixed.alpha = FALSE, nCores = 1)
2. estimate_convergent_regimes_surface(model, opt = opt)
3. cmp_model_score_CR(tr, Y, regimes, model$alpha, opt = opt)
4. cmp_AICc_CR(tree, Y, conv.regimes = regimes, alpha = alpha, opt = opt)
5. phylolm_interface_CR(tree, matrix(Y[, i]), conv.regimes, alpha = alpha[[i]], 
 .     opt = opt)
6. phylolm_CR(Y ~ preds - 1, phy = tr, model = "OUfixedRoot", sc = shift.configuration, 
 .     cr = conv.regimes, starting.value = alpha, upper.bound = opt$alpha.upper.bound, 
 .     lower.bound = alpha/100)
7. optim(logstart, fn = minus2llh, method = "L-BFGS-B", lower = loglower, 
 .     upper = logupper, ...)
8. (function (par) 
 . fn(par, ...))(-22.855602561775)
9. fn(par, ...)
10. loglik(prm, y, X = preds)
11. solve(comp$XX)
12. solve.default(comp$XX)
```

Code to reproduce:
```
library(l1ou)
tree = read.tree("tree.txt")
trait_matrix = read.table("traits.txt", sep="\t")
adj_data = adjust_data(tree=tree, Y=trait_matrix, normalize = FALSE)
fit_ind = estimate_shift_configuration(tree=adj_data$tree, Y=adj_data$Y, criterion="AICc", root.model="OUrandomRoot", nCores=4, rescale=TRUE)

# no error
fit_conv = estimate_convergent_regimes(fit_ind, criterion="AICc", method="backward", fixed.alpha=TRUE, nCores=4)

# error
fit_conv = estimate_convergent_regimes(fit_ind, criterion="AICc", method="backward", fixed.alpha=FALSE, nCores=4)

```

Many thanks.

Kenji

**Comment by khabbazian** [comment]
Hi Kenji,

Thank you for sending the error. I know what causes the error. I need some time to figure out how to fix it. So

If it is okay I will work on it more over the weekend and get back to you afterward.


Best regards,

Mohammad

________________________________

**Comment by kfuku52** [comment]
Thank you! I look forward to hearing your feedback. -Kenji

**Comment by khabbazian** [comment]
I updated the package with the fix. I couldn't run your code to make sure. Could you let me know if you still get the same error?

________________________________

**Comment by kfuku52** [comment]
Thank you. I updated the package and tried again but got another error in estimate_convergent_regimes().
```
Error in optim(logstart, fn = minus2llh, method = "L-BFGS-B", lower = loglower,  : 
  L-BFGS-B needs finite values of 'fn'
In addition: Warning message:
In mclapply(run.list, FUN = function(X) { :
  all scheduled cores encountered errors in user code
```

**Comment by khabbazian** [comment]
Could you please update the package and run your code again?

The problem is while optim tries to optimize the loglik, for the small values of alpha the design matrix becomes singular ( the condition number get less thatn .Machine$double.eps). I changed the tol in the solve(.). Let see if that helps.

________________________________

**Comment by cecileane** [comment]
Another idea, Mohammad, would be to call the log-likelihood function within a `try` or `tryCatch` block, and return a log-likelihood of `-Inf` if the design matrix becomes singular. That would force the estimate of α to be reasonably not too small.

**Comment by khabbazian** [comment]
I did something like that before this one but we got the following error

Error in optim(logstart, fn = minus2llh, method = "L-BFGS-B", lower = loglower, :

L-BFGS-B needs finite values of 'fn'



I will check if we get similar error if it returns NA.


________________________________

**Comment by kfuku52** [comment]
The updated version of estimate_convergent_regimes() is still running (>6 hrs from start, 16 cores allocated). I will get back to you when it finishes.

**Comment by khabbazian** [comment]
I guess because I changed the tol parameters of solve from 1e-16 to 1e-64 It became slower. Look forward.

________________________________

**Comment by kfuku52** [comment]
The process finished successfully without error. Thank you very much! Also, I would appreciate it if you could help fix [#29](https://github.com/khabbazian/l1ou/issues/29) when you have time.

**Comment by khabbazian** [comment]
Thank you. I will work on that too.

________________________________
